src/library/scala/concurrent/BatchingExecutor.scala

```diff
@@ -38,119 +43,129 @@ private[concurrent] object BatchingExecutorStatics {
+  */
+  private[this] final class Batch(private[this] final val resubmitOnBlock: Boolean) extends Runnable with BlockContext with (BlockContext => Throwable) {
+    private[this] final var parentBlockContext: BlockContext = BatchingExecutorStatics.MissingParentBlockContext
     private[this] final var first: Runnable = _
     private[this] final var size: Int = _
     private[this] final var other: Array[Runnable] = BatchingExecutorStatics.emptyBatchArray
 
-    def this(r: Runnable) = {
-      this()
+    def this(r: Runnable, resubmitOnBlock: Boolean) = {
+      this(resubmitOnBlock)
       first = r
       size = 1
     }
 
-    private def this(first: Runnable, other: Array[Runnable], size: Int) = {
-      this()
+    private def this(first: Runnable, other: Array[Runnable], size: Int, resubmitOnBlock: Boolean) = {
+      this(resubmitOnBlock)
       this.first = first
       this.other = other
       this.size = size
     }
 
     private[this] final def cloneAndClear(): Batch = {
-      val newBatch = new Batch(first, other, size)
+      val newBatch = new Batch(first, other, size, resubmitOnBlock)
       this.first = null
+      this.parentBlockContext = BatchingExecutorStatics.MissingParentBlockContext
       this.other = BatchingExecutorStatics.emptyBatchArray
       this.size = 0
       newBatch
     }
 
-    private[this] final def grow(): Unit = {
-      val len = other.length
-      other =
-        if (len == 0) new Array[Runnable](4)
-        else {
-          val newOther = new Array[Runnable](len << 1)
-          System.arraycopy(other, 0, newOther, 0, len)
-          newOther
-        }
+    private[this] final def ensureCapacity(curSize: Int): Array[Runnable] = {
+      val curOther = this.other
+      val curLen = curOther.length
+      if (curSize <= curLen) curOther
+      else {
+        val newLen = if (curLen == 0) 4 else curLen << 1
+
+        if (newLen <= curLen) throw new StackOverflowError("Space limit of asynchronous stack reached: " + curLen)
+        val newOther = new Array[Runnable](newLen)
+        System.arraycopy(curOther, 0, newOther, 0, curLen)
+        this.other = newOther
+        newOther
+      }
     }
 
     final def push(r: Runnable): Unit = {
-      val sz = size
-      if(sz > 0) {
-        if (sz > other.length)
-          grow()
-        other(sz - 1) = r
-      } else first = r
-      size = sz + 1
+      val sz = this.size
+      if(sz == 0)
+        this.first = r
+      else
+        ensureCapacity(sz)(sz - 1) = r
+      this.size = sz + 1
     }
 
-    final def pop(): Runnable =
-      (size: @switch) match {
-        case 0 => null
-        case 1 =>
-          val ret = first
-          first = null
-          size = 0
-          ret
-        case n =>
-          val ret = other(n - 2)
-          other(n - 2) = null
-          size = n - 1
-          ret
+    final def pop(): Runnable = {
+      val sz = this.size
```

Comment:
I'm not clearly understood, can you explain where is an improvement? It seems that additional instructions appears in case when size equals to 0.

Reply:
The rationale is that `pop()` is always invoked post a `size > 0`-check, so that means that the 0-case is never triggered. This change makes the method having a single branch instead of potentially 2, since we cannot know that the JVM will generate anything more efficient.

Reply:
the JVM would only have one branch ( because of @switch) ensuring that the bytecode was a `tableswitch` or `lookupswitch` etc (which generates a faster machine code as well on x64 than a series of `if` statements) , but an `if` is still probably faster here :-)

It may be worth adding a doc comment on the method here to explain that size > 0 should be checked by the caller

If size cannot be `< 1` should the test be `test == 1`?

Reply:
@mkeskells the size > 0 check is not mandatory, the logic will still work. AFAIK HotSpot passes both tableswitch and lookupswitch through the same codepath.

Reply:
@viktorklang they are different constraints that allow generation of different machine code

Reply:
@mkeskells Agreed. I've changed the code to not do the size-check before and instead do the tableswitch.